classes/that.sc

```diff
@@ -21,51 +21,54 @@ That {
 		all = ();
 	}
 
-	*new {|name, input, analyzerFunction, callback|
+	*new { |name, input, analyzerFunction, callback|
 		var res = all.at(name);
 		if(res.isNil, {
-			if(analyzerFunction.isNil, {
-				Error("Please provide an analyzerFunction").throw;
-			});
-			callback = callback ? {}; // enforce a function as callback and not allow nil
-			res = super.newCopyArgs(name, input, analyzerFunction, callback).init;
+			res = super.new.init;
 			all[name] = res;
-		}, {
-			input !? { res.input = input };
-			analyzerFunction !? { res.analyzerFunction = analyzerFunction };
-			callback !? { res.callback = callback };
 		});
+		res.setFunctions(input, analyzerFunction, callback);
 		^res
 	}
 
 	init {
 		defName = "that_%".format(name).asSymbol;
 		oscChannelName = "/that/%".format(name);
-		this.input = input;
+		analyzer = Ndef(defName);
+		oscdef = OSCdef(defName);
 	}
 
 	clear {
-		OSCdef(defName).free;
-		Ndef(defName).clear;
+		analyzer.free;
+		oscdef.clear;
 		all[name] = nil;
 	}
 
-	input_ {|newInput|
+	input_ { |newInput|
 		input = newInput;
-		this.prUpdateDefs();
+		this.prUpdateDefs;
 	}
 
-	analyzerFunction_ {|newAnalyzerFunction|
+	analyzerFunction_ { |newAnalyzerFunction|
 		analyzerFunction = newAnalyzerFunction;
-		this.prUpdateDefs();
+		this.prUpdateDefs;
+	}
+
+	setFunctions { |newInput, newAnalyzerFunction, newCallback|
```

Comment:
this should be a private function and should also be also excluded in the docs from showing up.

Also doesn't this crash if everything is not set/nil set during init?
I think we can default to an empty function `{}` for callback, but a `That` instance w/o an analyzer and input does not make sense to me and IMO it should be therefore impossible to create such a state as it leads to unintuitive interfaces? But maybe thats just my OCD from typesafety languages kicking in.

Reply:
> this should be a private function and should also be also excluded in the docs from showing up.

No, you want to be able to call this on an existing instance, too.

> Also doesn't this crash if everything is not set/nil set during init? I think we can default to an empty function `{}` for callback, 

It is typical in sclang to leave uninitialised values `nil`. The object `nil` responds to `value`, and one can always check with `notNil` to avoid building stuff. In proxies like `Ndef`, `nil` has also the meaning of a no-op and its always possible. The usage goes in the direction of a monoid.

>but a `That` instance w/o an analyzer and input does not make sense to me and IMO it should be therefore impossible to create such a state as it leads to unintuitive interfaces? But maybe thats just my OCD from typesafety languages kicking in.

The idea of a proxy like `That` is that you can use it before specifying it. So all the setters should be independent and atomic. You can e.g. also write `Ndef(\x, { Ndef(\y) + 2 })` without an `Ndef(\y)` in existence.

Reply:
> No, you want to be able to call this on an existing instance, too.

Well, its a matter of taste. I am more a Python person with tendency to Java in such regards, citing from [the zen of Python](https://www.python.org/dev/peps/pep-0020/#toc-entry-2)

> * There should be one-- and preferably only one --obvious way to do it.

which conflicts with the current implementation as it allows already to set the input/analyzer function in 3 ways and now one needs to maintain and test all 3 ways. But I know that the SC does not follow this rule and is more in opposition to this rule but in my experience as a programmer I found this one of the most fruitful rules for a nice to use library.

Regarding
> The idea of a proxy like That is that you can use it before specifying it.

well, I see this differently as I don't understand the state of such an object and for me it just makes things more complicated as it introduces ambiguous states (or uninitialized state) to an object instead of enforcing a state which always is *valid*.

In the end it comes down to preference I think.